### DISCUSSION TITLE
Ant+ multiplexing?

### DISCUSSION THREAD

**Discussion by talan-z** [doudar/SmartSpin2k#450] [Ideas]
Hi all,

I'm having a Schwinn Echolon Console that only supports ANT+. It would be fantastic if SmartSpin2k would allow to multiplex ANT+ to Bluetooth LE. 

I understand that the current board does not have ANT+. How much effort would it need to have ANT+ integrated?

Cheers,
talan

**Comment by doudar** [comment]
On the software side, I haven’t worked with ANT+ but I don’t believe it would be hard to implement.

On the hardware side, If there’s a microcontroller that can do BLE + WiFI + ANT+, we could support building for it easily but last I checked, there wasn’t an all-in-one solution so a co-processor would be needed such as the nrf-52. 

Overall, it’s probably not too hard of a project and I’d fully support implementing and integrating it, but it’s currently out of the scope I’m interested in working on directly.  I’m guessing the right person could have a prototype running in a weekend.

**Reply by talan-z** [reply]
Hi Anthony,

Thank you for sharing your thoughts! I am an absolute newbie when it comes to hardware design. Therefore unfortunately I won't be the guy who could have a prototype ready.  I wish there was an integrated microcontroller. I somehow thought that a USB dongle could be connected without having a significant hardware change.

Cheers,
talan

**Reply by doudar** [reply]
Unfortunately a USB connector wouldn't work. Like I say though - it wouldn't be very hard to add a NRF52 and have it communicate over I2C but it would take some development work.

**Reply by talan-z** [reply]
I ordered the [Nordic nRF52840 Dongle](https://www.nordicsemi.com/Products/Development-hardware/nrf52840-dongle) that has ANT capability. I should be able to build a simple ANT+ -> BLE bridge with it. Without any hardware experience, not sure whether I will be able to connect it directly to Spin2k. At first, I will try the bridge and also get some more experience with SmartSpin2k. If I understood correctly, I should be able to keep connecting my ANT+ devices directly to Zwift and the power control goes seperately via bluetooth to SmartSpin2k.

**Reply by MarkusSchneider** [reply]
With the Dongle you may have two options.
1) Connect the dongle to your PC an try to implement a program that takes BLE data, collected by your PC, and transform the data to ANT+ compliant data and send it via the dongle.
2) You can write your own code that can run on the dongle. But I'm not sure if the dongle can drive two communication channels at the same time. If this is possible you can collect the BLE data transform them and send them out via ANT+.

There is a commercial dongle available.
CABLE Connect is a dongle that transform the BLE data to ANT+ (unfortunately the opposite way)
https://npe-inc.com/cableinfo/
wyur is the replacement of cable connect www.npe-inc.com/wyur. Seems that this dongle can convert data in both ways.

**Reply by doudar** [reply]
What I would recomend (for development) is to write a program for the dongle that would send the ANT+ information to the SS2K (an esp32devkit) via i2c. If we had this, then possibly we could make R4 PCB with ANT+. I'm not even sure if it needs to be decoded - I think ANT+ and BLE data is very similar if not the same so we might be able to use the existing GAAT decoders built into SS2K already.

**Reply by doudar** [reply]
> I ordered the [Nordic nRF52840 Dongle](https://www.nordicsemi.com/Products/Development-hardware/nrf52840-dongle) that has ANT capability. I should be able to build a simple ANT+ -> BLE bridge with it. Without any hardware experience, not sure whether I will be able to connect it directly to Spin2k. At first, I will try the bridge and also get some more experience with SmartSpin2k. If I understood correctly, I should be able to keep connecting my ANT+ devices directly to Zwift and the power control goes seperately via bluetooth to SmartSpin2k.

Yes, if your device does both BLE and ANT+, you can use the ANT+ power directly to Zwift and send the BLE to SS2K so it can calculate ERG mode correctly.

**Reply by talan-z** [reply]
> There is a commercial dongle available. CABLE Connect is a dongle that transform the BLE data to ANT+ (unfortunately the opposite way) https://npe-inc.com/cableinfo/ wyur is the replacement of cable connect [www.npe-inc.com/wyur](http://www.npe-inc.com/wyur). Seems that this dongle can convert data in both ways.

Thanks, Markus. I am aware of these. Somehow these do not sell in my country and are quite pricey for my use case. I am still surprised that there's not open source ANT+ 2 BLE bridge anywhere in the net.



> What I would recomend (for development) is to write a program for the dongle that would send the ANT+ information to the SS2K (an esp32devkit) via i2c. If we had this, then possibly we could make R4 PCB with ANT+. I'm not even sure if it needs to be decoded - I think ANT+ and BLE data is very similar if not the same so we might be able to use the existing GAAT decoders built into SS2K already.

That's exactly the best case scenario what I am targeting. Let me check out the GAAT decoders.

**Comment by doudar** [comment]
> Hi all,
> 
> I'm having a Schwinn Echolon Console that only supports ANT+. It would be fantastic if SmartSpin2k would allow to multiplex ANT+ to Bluetooth LE.
> 
> I understand that the current board does not have ANT+. How much effort would it need to have ANT+ integrated?
> 
> Cheers, talan

Another option is....Do you think there's any chance the Schwinn bike communicates to the head unit via rs232 or i2c? In that case, it might be easier to actually just grab the data from there like we do with Peloton.

**Reply by talan-z** [reply]
Thank you for the suggestion! Actually, I haven't checked so far. An internet study did not reveal any hints.
It would definitely be a less preferred option as I do not want to "decipher" their data and put it into BLE. I feel the ANT+ receiver/bridge would be the most elegant way
